src/allotropy/allotrope/schema_parser/schema_cleaner.py

```diff
@@ -0,0 +1,621 @@
+        # dereference defintion references. Though this smells, the flag is nested so deep in the
+        # code, that it would be impractical to send this all the way down, and I think this is the
+        # lesser evil.
+        self.enclosing_schema_name: Optional[str] = None
+        self.enclosing_schema_keys: Optional[dict[str, Any]] = None
+
+    def get_referenced_units(self) -> dict[str, str]:
+        # Returns all units seen in the schema, for auto-generating unit schemas externally.
+        return {
+            unit: iri
+            for unit, iri in self.unit_to_iri.items()
+            if unit in self.referenced_units
+        }
+
+    def _is_unit_name_ref(self, ref: str) -> bool:
+        return _get_def_name(ref) in self.unit_to_name.values()
+
+    def _add_unit(self, unit: str, unit_iri: str) -> None:
+        self.unit_to_name[unit] = unit_name_from_iri(unit_iri)
+        self.unit_to_iri[unit] = unit_iri
+
+    def _is_quantity_value(self, values: list[dict[str, Any]]) -> bool:
+        # Check if this schema is a special case of allOf: [tQuantityValue, unit] and if so replace.
+        if len(values) != 2:  # noqa: PLR2004
+            return False
+        quantity_value_found = False
+        unit_found = False
+        for value in values:
+            if not _is_ref_schema(value):
+                return False
+            if _get_def_name(value["$ref"]) == "tQuantityValue":
+                quantity_value_found = True
+            if self._is_unit_name_ref(value["$ref"]):
+                unit_found = True
+        return quantity_value_found and unit_found
+
+    def _fix_quantity_value_reference(
+        self, values: list[dict[str, Any]]
+    ) -> dict[str, Any]:
+        for value in values:
+            def_name = _get_def_name(value["$ref"])
+            if def_name == "tQuantityValue":
+                continue
+            return {"$ref": f"#/$defs/tQuantityValue{def_name}"}
+
+        msg = f"Failed to find value unit in quantity value reference: {values}"
+        raise AssertionError(msg)
+
+    def _try_combine_object_schemas(
+        self, schemas: list[dict[str, Any]]
+    ) -> dict[str, Any]:
+        # Try to combine a flattened list of object schemas
+        if any(_is_array_schema(schema) for schema in schemas):
+            msg = "Unexpected array schema in _try_combine_object_schemas"
+            raise AssertionError(msg)
+
+        if any(_is_composed_object_schema(schema) for schema in schemas):
+            msg = "Unexpected composed object schema in _try_combine_object_schemas"
+            raise AssertionError(msg)
+
+        all_values = defaultdict(list)
+        for schema in schemas:
+            for key, value in schema.get("properties", {}).items():
+                all_values[key].append(value)
+
+        combined_props = {}
+        for key, values in dict(all_values).items():
+            if len(values) == 1 or _all_values_equal(values):
```

Comment:
Yeah, added as an optimization, but (1) I don't know that it wouldn't be optimized away (2) even if so, it would belong in the helper. Moved it to the helper, we can let the compiler decide if it wants to keep it.